### PR TITLE
add program id validation to clinical upload

### PR DESCRIPTION
**Pull request by blabadi**
stream records to process different validations
move stuff around

**Description of changes**

**Type of Change**

- [ ] Bug
- [ ] Refactor
- [ ] New Feature

**Checklist before requesting review:**

- [ ] Check branch (code change PRs go to `develop` not master)
- [ ] Manual testing
- [ ] Regression tests completed and passing (double check number of tests).
- [ ] Spelling has been checked.
- [ ] Updated swagger docs accordingly (check it's still valid)

### CODE REVIEW

File: src/submission/validation-clinical/validation.ts [hunk 1]
```diff
@@ -0,0 +1,623 @@
+import { DonorMap, Specimen, Sample, Donor } from '../../clinical/clinical-entities';
```
Comment (blabadi) [inline src/submission/validation-clinical/validation.ts]: I moved this file from /submission to /validation-clinical .. no new content

File: src/submission/submission-service.ts [hunk 2]
```diff
@@ -90,43 +93,44 @@ export namespace operations {
...
-    }
-
+    let validRecordsAccumulator: DeepReadonly<any[]> = [];
     // check the program id if it matches the authorized one
     // This check is used to validate the program Id along with the schema validations
     // to save extra round trips
-    let programIdErrors: DeepReadonly<SubmissionValidationError[]> = [];
     command.records.forEach((r, index) => {
-      const programIdError = dataValidator.usingInvalidProgramId(
+      const schemaResult = schemaManager
```
Comment (blabadi) [inline src/submission/submission-service.ts]: instead of two iterations over data, do one iteration

File: src/lectern-client/schema-functions.ts [hunk 3]
```diff
@@ -57,34 +57,72 @@ export const process = (
...
+    `done processing all rows, validationErrors: ${validationErrors.length}, validRecords: ${processedRecords.length}`,
+  );
+
+  return F({
+    validationErrors,
+    processedRecords,
+  });
+};
+
+export const process = (
```
Comment (blabadi) [inline src/lectern-client/schema-functions.ts]: a new method to process one record instead of batch but keeps index information

Comment (jaserud) [reply src/lectern-client/schema-functions.ts]: cool, but it looks a little weird, this function is processing one record, but it returns `prcessedRecords` which is an array that only contains the one record?? guess it makes it easy if both functions return arrays... Is that what you were going for?

File: src/submission/submission-service.ts [hunk 4]
```diff
@@ -90,43 +93,44 @@ export namespace operations {
...
+      const programIdError = usingInvalidProgramId(
         ClinicalEntityType.REGISTRATION,
         index,
         r,
         command.programId,
       );
-      programIdErrors = programIdErrors.concat(programIdError);
+
+      unifiedSchemaErrors = unifiedSchemaErrors.concat(programIdError);
+      validRecordsAccumulator = validRecordsAccumulator.concat(schemaResult.processedRecords);
```
Comment (jaserud) [inline src/submission/submission-service.ts]: same comment as earlier...

File: src/submission/submission-service.ts [hunk 5]
```diff
@@ -504,6 +508,7 @@ export namespace operations {
     const validateResult: ClinicalTypeValidateResult = await validateSubmissionData(
       clinicalSubmissionRecords,
       relevantDonorsMap,
+      command.programId,
```
Comment (jaserud) [inline src/submission/submission-service.ts]: why do we need to pass `command.programId` now, does it somehow use it??

File: src/lectern-client/schema-functions.ts (hunk 3 above)
Comment (blabadi) [reply src/lectern-client/schema-functions.ts]: I'll take a look .. I don't think I left it like that for a reason

File: src/submission/submission-service.ts (hunk 5 above)
Comment (blabadi) [reply src/submission/submission-service.ts]: Nope I think I added it here first before moving the programId validation logic to upload

**Pull request merged**